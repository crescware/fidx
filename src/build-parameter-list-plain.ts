import type { FunctionInfo } from "./function-info";

export function buildParameterListPlain(func: FunctionInfo): string {
	const formattedParameters = func.parameters
		.map((param) => {
			const name = param.name;
			const type = param.type ? `: ${param.type}` : "";
			return `${name}${type}`;
		})
		.join(", ");

	return formattedParameters ? `(${formattedParameters})` : "()";
}
