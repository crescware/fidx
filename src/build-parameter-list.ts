import { dim } from "yoctocolors";

import type { FunctionInfo } from "./function-info";
import { typeMetadata } from "./type-metadata";

export function buildParameterList(func: FunctionInfo): string {
	const formattedParameters = func.parameters
		.map((param) => {
			const name = typeMetadata[func.type].color(param.name);
			const type = param.type
				? `${dim(":")} ${typeMetadata[func.type].color(param.type)}`
				: "";
			return `${name}${type}`;
		})
		.join(`${dim(",")} `);

	return formattedParameters
		? `${dim("(")}${formattedParameters}${dim(")")}`
		: `${dim("()")}`;
}
