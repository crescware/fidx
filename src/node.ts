import type { walk } from "oxc-walker";

type WalkCallback = Parameters<typeof walk>[1];
export type Node = Parameters<NonNullable<WalkCallback["enter"]>>[0];
