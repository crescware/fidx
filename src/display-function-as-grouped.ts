import { relative } from "node:path";
import { dim } from "yoctocolors";

import type { extractAllFunctions } from "./extract-all-functions";
import { makeColoredName } from "./make-colored-name";
import { typeMetadata } from "./type-metadata";

export function displayFunctionAsGrouped(
	targetDir: string,
	results: Awaited<ReturnType<typeof extractAllFunctions>>,
	useAbsolutePaths: boolean,
): { totalFunctions: number; fileCount: number } {
	let totalFunctions = 0;
	let fileCount = 0;

	const maxLineNumber = Math.max(
		...results.flatMap((r) => r.functions.map((f) => f.line)),
	);
	const lineNumberWidth = String(maxLineNumber).length;

	for (const result of results) {
		const path = useAbsolutePaths
			? result.filePath
			: relative(targetDir, result.filePath);
		console.log(dim(path));

		for (const func of result.functions) {
			const colored = makeColoredName(func.name, func.type);
			const paddedLineNumber = dim(
				String(func.line).padStart(lineNumberWidth, " "),
			);
			const returnType = func.returnType
				? `${dim(":")} ${typeMetadata[func.type].color(func.returnType)}`
				: "";
			console.log(`  ${paddedLineNumber} ${colored}${returnType}`);
			totalFunctions++;
		}

		console.log();
		fileCount++;
	}

	return { totalFunctions, fileCount };
}
