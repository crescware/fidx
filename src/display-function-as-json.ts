import { relative } from "node:path";
import { JsonStreamStringify } from "json-stream-stringify";

import type { extractAllFunctions } from "./extract-all-functions";

export async function displayFunctionAsJson(
	targetDir: string,
	results: Awaited<ReturnType<typeof extractAllFunctions>>,
	useAbsolutePaths: boolean,
): Promise<void> {
	const output = results.map((result) => {
		const path = useAbsolutePaths
			? result.filePath
			: relative(targetDir, result.filePath);

		return {
			path,
			functions: result.functions.map((v) => ({
				line: v.line,
				name: v.name,
				returnType: v.returnType,
			})),
		};
	});

	const stream = new JsonStreamStringify(output);

	return new Promise<void>((resolve, reject) => {
		stream.on("data", (v) => process.stdout.write(v));
		stream.on("error", (e) => reject(e));

		stream.on("end", () => {
			process.stdout.write("\n");
			resolve();
		});
	});
}
