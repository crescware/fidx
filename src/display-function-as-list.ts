import { relative } from "node:path";
import { dim } from "yoctocolors";

import type { extractAllFunctions } from "./extract-all-functions";
import { makeColoredName } from "./make-colored-name";

export function displayFunctionAsList(
	targetDir: string,
	results: Awaited<ReturnType<typeof extractAllFunctions>>,
	useAbsolutePaths: boolean,
): { totalFunctions: number; fileCount: number } {
	let totalFunctions = 0;
	let fileCount = 0;

	for (const result of results) {
		const path = useAbsolutePaths
			? result.filePath
			: relative(targetDir, result.filePath);

		for (const func of result.functions) {
			const location = `${path}:${func.line}`;
			const colored = makeColoredName(func.name, func.type);
			console.log(`${dim(location)} ${colored}`);
			totalFunctions++;
		}

		fileCount++;
	}

	return { totalFunctions, fileCount };
}
