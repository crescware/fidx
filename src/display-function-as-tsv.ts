import { relative } from "node:path";

import { buildParameterListPlain } from "./build-parameter-list-plain";
import type { extractAllFunctions } from "./extract-all-functions";

export function displayFunctionAsTsv(
	targetDir: string,
	results: Awaited<ReturnType<typeof extractAllFunctions>>,
	useAbsolutePaths: boolean,
): void {
	const headers = ["path", "line", "name", "parameters", "returnType"];
	console.log(headers.join("\t"));

	for (const result of results) {
		const path = useAbsolutePaths
			? result.filePath
			: relative(targetDir, result.filePath);

		for (const func of result.functions) {
			const parameterList = buildParameterListPlain(func);
			const row = [
				path,
				func.line,
				func.name,
				parameterList,
				func.returnType || "",
			];
			console.log(row.join("\t"));
		}
	}
}
