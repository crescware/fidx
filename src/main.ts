import { resolve } from "node:path";
import arg from "arg";

import { displayFunctionExtractionResults } from "./display-function-extraction-results";
import { displayFunctionExtractionResultsAsList } from "./display-function-extraction-results-as-list";
import { displaySummary } from "./display-summary";
import { divider } from "./divider";
import { extractAllFunctions } from "./extract-all-functions";
import { PreconditionError } from "./precondition-error";

export async function main(): Promise<void> {
	const args = arg({
		"--report": String,
	});

	const targetPath = args._[0];
	if (!targetPath) {
		throw new PreconditionError("Directory path is required");
	}
	const targetDir = resolve(process.env.INIT_CWD || ".", targetPath);
	const startTime = Date.now();
	const results = await extractAllFunctions(targetDir);
	const endTime = Date.now();
	const elapsedSeconds = (endTime - startTime) / 1000;

	console.log("Function extraction complete!\n");
	console.log(divider());

	const reportType = args["--report"] || "table";

	if (reportType === "table") {
		const { totalFunctions, fileCount } = displayFunctionExtractionResults(
			targetDir,
			results,
		);

		displaySummary(totalFunctions, fileCount, elapsedSeconds);
		return;
	}

	if (reportType === "list") {
		const { totalFunctions, fileCount } =
			displayFunctionExtractionResultsAsList(targetDir, results);

		displaySummary(totalFunctions, fileCount, elapsedSeconds);
		return;
	}

	throw new PreconditionError("invalid config");
}
