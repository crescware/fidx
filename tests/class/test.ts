import { resolve } from "node:path";
import {
	afterAll,
	afterEach,
	beforeEach,
	describe,
	expect,
	test,
	vi,
} from "vitest";
import { main } from "../../src/main";

describe("class methods detection", () => {
	const classFixtureDir = resolve(import.meta.dirname, "src");
	const originalArgv = process.argv;
	const originalConsoleLog = console.log;
	const originalStdoutWrite = process.stdout.write;
	let consoleOutput: string[] = [];

	beforeEach(() => {
		consoleOutput = [];
		console.log = vi.fn((...args) => {
			consoleOutput.push(args.join(" "));
		});
		process.stdout.write = vi.fn((chunk) => {
			consoleOutput.push(chunk.toString());
			return true;
		});

		let mockTime = 0;
		vi.spyOn(Date, "now").mockImplementation(() => {
			mockTime += 1000;
			return mockTime;
		});

		process.argv = ["node", "fidx", classFixtureDir, "--format", "grouped"];
	});

	afterEach(() => {
		process.argv = originalArgv;
		console.log = originalConsoleLog;
		process.stdout.write = originalStdoutWrite;
		vi.clearAllMocks();
	});

	afterAll(() => {
		vi.restoreAllMocks();
	});

	describe("grouped format", () => {
		test("should detect constructor and methods", async () => {
			await main();
			const output = consoleOutput.join("\n");

			expect(output).toContain("class.ts");
			// Constructor should be yellow
			expect(output).toContain(
				"\x1b[2m 2\x1b[22m \x1b[33mconstructor\x1b[39m\x1b[2m(\x1b[22m\x1b[33ma\x1b[39m\x1b[2m:\x1b[22m \x1b[33mstring\x1b[39m\x1b[2m,\x1b[22m \x1b[33mb\x1b[39m\x1b[2m:\x1b[22m \x1b[33mnumber\x1b[39m\x1b[2m)\x1b[22m",
			);
			// Methods should be yellow
			expect(output).toContain(
				"\x1b[2m 7\x1b[22m \x1b[33mmethodOne\x1b[39m\x1b[2m(\x1b[22m\x1b[33mx\x1b[39m\x1b[2m:\x1b[22m \x1b[33mstring\x1b[39m\x1b[2m)\x1b[22m\x1b[2m:\x1b[22m \x1b[33mvoid\x1b[39m",
			);
			expect(output).toContain(
				"\x1b[2m11\x1b[22m \x1b[33mmethodTwo\x1b[39m\x1b[2m(\x1b[22m\x1b[33my\x1b[39m\x1b[2m:\x1b[22m \x1b[33mnumber\x1b[39m\x1b[2m,\x1b[22m \x1b[33mz\x1b[39m\x1b[2m:\x1b[22m \x1b[33mboolean\x1b[39m\x1b[2m)\x1b[22m\x1b[2m:\x1b[22m \x1b[33mstring\x1b[39m",
			);
			expect(output).toContain(
				"\x1b[2m17\x1b[22m \x1b[33mstaticMethod\x1b[39m\x1b[2m(\x1b[22m\x1b[33mparam\x1b[39m\x1b[2m:\x1b[22m \x1b[33mstring[]\x1b[39m\x1b[2m)\x1b[22m\x1b[2m:\x1b[22m \x1b[33mnumber\x1b[39m",
			);
		});

		test("should count methods correctly", async () => {
			await main();
			const output = consoleOutput.join("\n");
			expect(output).toContain("Total: 4 functions found in 1 files");
		});

		test("should match snapshot", async () => {
			await main();
			const output = consoleOutput.join("\n");
			expect(output).toMatchSnapshot();
		});
	});

	describe("list format", () => {
		test("should output class methods in list format", async () => {
			process.argv = ["node", "fidx", classFixtureDir, "--format", "list"];
			await main();
			const output = consoleOutput.join("\n");
			expect(output).toContain("class.ts:2");
			expect(output).toContain("constructor");
			expect(output).toContain("class.ts:7");
			expect(output).toContain("methodOne");
			expect(output).toContain("class.ts:11");
			expect(output).toContain("methodTwo");
			expect(output).toContain("class.ts:17");
			expect(output).toContain("staticMethod");
		});
	});

	describe("tsv format", () => {
		test("should output class methods in TSV format", async () => {
			process.argv = ["node", "fidx", classFixtureDir, "--format", "tsv"];
			await main();
			const output = consoleOutput.join("\n");
			expect(output).toContain("path\tline\tname");
			expect(output).toContain("class.ts\t2\tconstructor");
			expect(output).toContain("class.ts\t7\tmethodOne");
			expect(output).toContain("class.ts\t11\tmethodTwo");
			expect(output).toContain("class.ts\t17\tstaticMethod");
		});
	});

	describe("json format", () => {
		test("should output class methods in JSON format with parameters and return types", async () => {
			process.argv = ["node", "fidx", classFixtureDir, "--format", "json"];
			await main();
			const output = consoleOutput.join("");
			expect(output).toContain('[{"path":"class.ts","functions":[');
			expect(output).toContain(
				'{"line":2,"name":"constructor","parameters":[{"name":"a","type":"string"},{"name":"b","type":"number"}],"returnType":null}',
			);
			expect(output).toContain(
				'{"line":7,"name":"methodOne","parameters":[{"name":"x","type":"string"}],"returnType":"void"}',
			);
			expect(output).toContain(
				'{"line":11,"name":"methodTwo","parameters":[{"name":"y","type":"number"},{"name":"z","type":"boolean"}],"returnType":"string"}',
			);
			expect(output).toContain(
				'{"line":17,"name":"staticMethod","parameters":[{"name":"param","type":"string[]"}],"returnType":"number"}',
			);
		});
	});
});
