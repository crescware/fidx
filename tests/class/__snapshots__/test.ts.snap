// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`class methods detection > grouped format > should match snapshot 1`] = `
"Extracting functions from TypeScript files...

Glob execution time: 1s
Parse time: 1s
Function extraction complete!

============================================================

[2mLegend:[22m
  [36mname()[39m - Functions and arrow functions
  [33mname()[39m - Methods and constructors

[2mclass.ts[22m
  [2m 2[22m [33mconstructor[39m[2m([22m[33ma[39m[2m:[22m [33mstring[39m[2m,[22m [33mb[39m[2m:[22m [33mnumber[39m[2m)[22m
  [2m 7[22m [33mmethodOne[39m[2m([22m[33mx[39m[2m:[22m [33mstring[39m[2m)[22m[2m:[22m [33mvoid[39m
  [2m11[22m [33mmethodTwo[39m[2m([22m[33my[39m[2m:[22m [33mnumber[39m[2m,[22m [33mz[39m[2m:[22m [33mboolean[39m[2m)[22m[2m:[22m [33mstring[39m
  [2m17[22m [33mstaticMethod[39m[2m([22m[33mparam[39m[2m:[22m [33mstring[][39m[2m)[22m[2m:[22m [33mnumber[39m

============================================================


Total: 4 functions found in 1 files
Time: 5.00 seconds
"
`;
